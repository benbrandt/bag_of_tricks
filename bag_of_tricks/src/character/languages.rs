use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Character;

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Auran,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub(crate) fn gen(rng: &mut impl Rng, character: &Character, amount: usize) -> Vec<Self> {
        Language::iter()
            .filter(|l| !character.languages.contains(l))
            .choose_multiple(rng, amount)
    }
}

/// Trait for describing what languages are provided by a given object.
pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> usize {
        0
    }
}
