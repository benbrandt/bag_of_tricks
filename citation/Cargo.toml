[package]
name = "citation"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2018"
resolver = "2"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
serde = { version = "1", features = ["derive"] }
strum = { version = "0.20", features = ["derive"] }
