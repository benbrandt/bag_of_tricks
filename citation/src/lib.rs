#![deny(clippy::all)]
#![warn(clippy::pedantic)]
use std::{
    collections::{HashMap, HashSet},
    fmt,
};

use serde::{Deserialize, Serialize};
use strum::Display;

/// Titles of the available D&D Books.
#[derive(Clone, Copy, Debug, Deserialize, Display, Eq, Hash, PartialEq, Serialize)]
pub enum Book {
    #[strum(serialize = "PHB")]
    Phb,
    #[strum(serialize = "SCAG")]
    Scag,
    #[strum(serialize = "VGTM")]
    Vgtm,
}

/// Book and page number for citations.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct Citation(pub Book, pub u16);

impl fmt::Display for Citation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} p{}", self.0, self.1)
    }
}

/// List of multiple citations.
#[derive(Deserialize, Serialize)]
pub struct CitationList(pub Vec<Citation>);

/// Displays multiple citations from the same book together.
impl fmt::Display for CitationList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut citations = HashMap::new();
        for c in &self.0 {
            citations
                .entry(c.0)
                .or_insert_with(HashSet::new)
                .insert(c.1);
        }
        let mut books = citations
            .iter()
            .map(|(book, page_nums)| {
                let mut pages = page_nums
                    .iter()
                    .map(ToString::to_string)
                    .collect::<Vec<_>>();
                pages.sort();
                format!("{} p{}", book, pages.join(","))
            })
            .collect::<Vec<_>>();
        books.sort();
        write!(f, "{}", books.join(" "))
    }
}

/// Trait for any entity/object in need of citation.
///
/// Makes it easer for users to find more information in the source books.
pub trait Citations {
    /// Return list of citations for the object in question
    fn citations(&self) -> CitationList;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_book_display() {
        assert_eq!(format!("{}", Book::Phb), "PHB");
    }

    #[test]
    fn test_citation_display() {
        let citation = Citation(Book::Phb, 123);
        assert_eq!(format!("{}", citation), "PHB p123");
    }

    #[test]
    fn test_citations_display() {
        let citations = CitationList(vec![
            Citation(Book::Phb, 123),
            Citation(Book::Phb, 125),
            Citation(Book::Vgtm, 125),
        ]);
        assert_eq!(format!("{}", citations), "PHB p123,125 VGTM p125");
    }
}
