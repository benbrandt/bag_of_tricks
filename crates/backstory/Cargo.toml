[package]
name = "backstory"
version = "0.1.0"
authors = ["Ben Brandt <benjamin.j.brandt@gmail.com>"]
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
