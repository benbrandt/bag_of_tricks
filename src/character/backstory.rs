pub(crate) trait Backstory {
    fn backstory(&self) -> Vec<String> {
        vec![]
    }
}
