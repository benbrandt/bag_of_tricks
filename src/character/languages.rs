use serde::{Deserialize, Serialize};
use strum_macros::{Display, EnumIter};

#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Does this provide additional language options?
    fn addl_languages(&self) -> usize {
        0
    }
}
