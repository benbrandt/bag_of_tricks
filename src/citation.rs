use serde::{Deserialize, Serialize};
use std::{
    collections::{HashMap, HashSet},
    fmt,
};
use strum_macros::Display;

/// Titles of the available D&D Books.
#[derive(Clone, Copy, Debug, Deserialize, Display, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Book {
    #[strum(serialize = "PHB")]
    Phb,
    #[strum(serialize = "VGTM")]
    Vgtm,
}

/// Book and page number for citations.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub(crate) struct Citation(pub(crate) Book, pub(crate) u16);

impl fmt::Display for Citation {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} p{}", self.0, self.1)
    }
}

/// List of multiple citations.
#[derive(Deserialize, Serialize)]
pub(crate) struct CitationList(pub(crate) Vec<Citation>);

/// Displays multiple citations from the same book together.
impl fmt::Display for CitationList {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let mut citations = HashMap::new();
        for c in &self.0 {
            citations
                .entry(c.0)
                .or_insert_with(HashSet::new)
                .insert(c.1);
        }
        for (book, page_nums) in citations {
            let mut pages = page_nums
                .iter()
                .map(ToString::to_string)
                .collect::<Vec<_>>();
            pages.sort();
            write!(f, "{} p{}", book, pages.join(","))?;
        }
        write!(f, "")
    }
}

/// Trait for any entity/object in need of citation.
///
/// Makes it easer for users to find more information in the source books.
pub(crate) trait Citations {
    /// Return list of citations for the object in question
    fn citations(&self) -> CitationList;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_book_display() {
        assert_eq!(format!("{}", Book::Phb), "PHB");
    }

    #[test]
    fn test_citation_display() {
        let citation = Citation(Book::Phb, 123);
        assert_eq!(format!("{}", citation), "PHB p123");
    }

    #[test]
    fn test_citations_display() {
        let citations = CitationList(vec![Citation(Book::Phb, 123), Citation(Book::Phb, 125)]);
        assert_eq!(format!("{}", citations), "PHB p123,125");
    }
}
