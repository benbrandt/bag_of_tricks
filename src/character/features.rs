use serde::{Deserialize, Serialize};
use std::fmt;

use crate::citation::Citation;

/// A feature or trait a character has.
#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub(crate) struct Feature {
    /// Name of the feature or trait.
    pub(crate) title: &'static str,
    /// Citation for where more information about this feature is available.
    pub(crate) citation: Citation,
}

impl fmt::Display for Feature {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} ({})", self.title, self.citation)
    }
}

/// Trait for objects that provide features to a character.
pub(crate) trait Features {
    // Return a list of features this thing provides
    fn features(&self) -> Vec<Feature> {
        vec![]
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::citation::Book;

    #[test]
    fn test_display() {
        let feature = Feature {
            title: "Title",
            citation: Citation(Book::Phb, 1),
        };
        insta::assert_snapshot!(format!("{}", feature));
    }
}
