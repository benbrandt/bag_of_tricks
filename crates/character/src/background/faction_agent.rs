use std::fmt;

use backstory::Backstory;
use citation::{Book, Citation, CitationList, Citations};
use features::{Feature, Features};
use gear::{
    adventuring_gear::{Gear, OtherGear},
    currency::Coin,
};
use languages::{LanguageType, Languages};
use personality::{Influence, PersonalityOptions};
use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use stats::{
    ability::{AbilityScoreType, AbilityScores, Skill},
    equipment::{Equipment, EquipmentOption, StartingEquipment},
    proficiencies::{Proficiencies, Proficiency, ProficiencyOption},
};
use strum::{Display, EnumIter, IntoEnumIterator};

use super::Background;

const SKILLS: &[Skill] = &[Skill::Insight];
const BONDS: &[&str] = &[
    "I would die to recover an ancient relic of my faction that was lost long ago.",
    "I will someday get revenge on the corrupt faction hierarchy who branded me a traitor.",
    "I owe my life to the faction member who took me in when my parents died.",
    "Everything I do is for the common people.",
    "I will do anything to protect the network where I served.",
    "I seek to preserve a secret text that my enemies consider heretical and seek to destroy.",
];
const FLAWS: &[&str] = &[
    "I judge others harshly, and myself even more severely.",
    "I put too much trust in those who wield power within my faction's hierarchy.",
    "My loyalty sometimes leads me to blindly trust those that profess membership in my faction.",
    "I am inflexible in my thinking.",
    "I am suspicious of strangers and expect the worst of them.",
    "Once I pick a goal, I become obsessed with it to the detriment of everything else in my life.",
];
const IDEALS: &[(&str, Influence)] = &[
    ("Tradition. The ancient traditions of membership and secrecy must be preserved and upheld.", Influence::Lawful),
    ("Charity. I always try to help those in need, no matter what the personal cost.", Influence::Good),
    ("Change. We must help bring about the changes the faction is constantly working in the world", Influence::Chaotic),
    ("Power. I hope to one day rise to the top of my faction's hierarchy.", Influence::Lawful),
    ("Faith. I trust that my faction will guide my actions. I have faith that if I work hard, things will go well.", Influence::Lawful),
    ("Aspiration. I seek to prove myself worthy of my faction's favor by matching my actions against their teachings.", Influence::Any),
];
const TRAITS: &[&str] = &[
    "I idolize a particular hero of my faction, and constantly refer to that person's deeds and example.",
    "I can find common ground between the fiercest enemies, empathizing with them and always working toward peace.",
    "I see omens in every event and action. The gods try to speak to us, we just need to listen.",
    "Nothing can shake my optimistic attitude.",
    "I quote (or misquote) faction texts and proverbs in almost every situation.",
    "I am tolerant (or intolerant) of other factions and respect (or condemn) the membership in other factions.",
    "I've enjoyed fine food, drink, and high society among my faction's elite. Rough living grates on me.",
    "I've spent so long in the faction that I have little practical experience dealing with people in the outside world.",
];

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Faction {
    #[strum(serialize = "The Harpers")]
    Harpers,
    #[strum(serialize = "The Order of the Gauntlet")]
    OrderOfTheGauntlet,
    #[strum(serialize = "The Emerald Enclave")]
    EmeraldEnclave,
    #[strum(serialize = "The Lord's Alliance")]
    LordsAlliance,
    #[strum(serialize = "The Zhentarim")]
    Zhentarim,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct FactionAgent {
    faction: Faction,
}

impl FactionAgent {
    fn addl_skills() -> Vec<Skill> {
        Skill::iter()
            .filter(|s| {
                [
                    AbilityScoreType::Charisma,
                    AbilityScoreType::Intelligence,
                    AbilityScoreType::Wisdom,
                ]
                .contains(&s.ability_score_type())
            })
            .collect::<Vec<_>>()
    }
}

#[typetag::serde]
impl Background for FactionAgent {
    fn gen(
        rng: &mut impl Rng,
        _: &AbilityScores,
        _: &[Proficiency],
        _: i16,
    ) -> Box<dyn Background> {
        Box::new(Self {
            faction: Faction::iter().choose(rng).unwrap(),
        })
    }

    fn skills() -> Vec<Skill> {
        let mut skills = SKILLS.to_vec();
        skills.extend(Self::addl_skills());
        skills
    }
}

impl Backstory for FactionAgent {}

impl Citations for FactionAgent {
    fn citations(&self) -> CitationList {
        CitationList(vec![Citation(Book::Scag, 147)])
    }
}

impl Features for FactionAgent {
    fn features(&self) -> Vec<Feature> {
        // As a faction agent, you have access to a secret network of supporters and operatives who can provide assistance on your adventures. You know a set of secret signs and passwords you can use to identify such operatives, who can provide you with access to a hidden safe house, free room and board, or assistance in finding information. These agents never risk their lives for you or risk revealing their true identities.
        vec![Feature {
            title: "Safe Haven",
            citation: Citation(Book::Scag, 147),
        }]
    }
}

impl Languages for FactionAgent {
    fn addl_languages(&self) -> (usize, Option<LanguageType>) {
        (2, None)
    }
}

impl PersonalityOptions for FactionAgent {
    fn bonds(&self) -> Vec<String> {
        BONDS.iter().map(|&s| s.to_string()).collect()
    }

    fn flaws(&self) -> Vec<String> {
        FLAWS.iter().map(|&s| s.to_string()).collect()
    }

    fn ideals(&self) -> Vec<(String, Influence)> {
        IDEALS.iter().map(|&(s, i)| (s.to_string(), i)).collect()
    }

    fn traits(&self) -> Vec<String> {
        TRAITS.iter().map(|&s| s.to_string()).collect()
    }
}

impl Proficiencies for FactionAgent {
    fn proficiencies(&self) -> Vec<Proficiency> {
        SKILLS.iter().map(|&s| Proficiency::Skill(s)).collect()
    }

    fn addl_proficiencies(&self) -> Vec<ProficiencyOption> {
        vec![ProficiencyOption::Skill(Some(Self::addl_skills()), 1)]
    }
}

impl StartingEquipment for FactionAgent {
    fn coins(&self) -> (Coin, u8) {
        (Coin::Gold, 15)
    }

    fn equipment(&self) -> Vec<Equipment> {
        vec![
            Equipment::Gear(Gear::Other(OtherGear::ClothesCommon)),
            Equipment::Gear(Gear::Other(OtherGear::Pouch)),
        ]
    }

    fn addl_equipment(&self) -> Vec<EquipmentOption> {
        vec![
            EquipmentOption::From(
                ["badge", "emblem"]
                    .iter()
                    .map(|i| Equipment::Other(format!("the {} of your faction", i)))
                    .collect(),
            ),
            EquipmentOption::From(
                [
                    "a copy of a seminal faction text",
                    "a code-book for a covert faction",
                ]
                .iter()
                .map(|&i| Equipment::Other(i.to_string()))
                .collect(),
            ),
        ]
    }
}

impl fmt::Display for FactionAgent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} Faction Agent", self.faction)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background);
    }

    #[test]
    fn test_snapshot_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_display_snapshot!(background);
    }

    #[test]
    fn test_bonds() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.bonds());
    }

    #[test]
    fn test_flaws() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.flaws());
    }

    #[test]
    fn test_ideals() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.ideals());
    }

    #[test]
    fn test_traits() {
        let mut rng = Pcg64::seed_from_u64(1);
        let background = FactionAgent::gen(&mut rng, &AbilityScores::default(), &[], 2);
        insta::assert_yaml_snapshot!(background.traits());
    }

    #[test]
    fn test_skills() {
        insta::assert_yaml_snapshot!(FactionAgent::skills());
    }

    #[test]
    fn test_snapshot_citations() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.citations());
    }

    #[test]
    fn test_snapshot_features() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.features());
    }

    #[test]
    fn test_snapshot_languages() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.addl_languages());
    }

    #[test]
    fn test_snapshot_proficiences() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.proficiencies());
    }

    #[test]
    fn test_snapshot_addl_proficiences() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.addl_proficiencies());
    }

    #[test]
    fn test_snapshot_coins() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.coins());
    }

    #[test]
    fn test_snapshot_equipment() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.equipment());
    }

    #[test]
    fn test_snapshot_addl_equipment() {
        let background = FactionAgent {
            faction: Faction::EmeraldEnclave,
        };
        insta::assert_yaml_snapshot!(background.addl_equipment());
    }
}
