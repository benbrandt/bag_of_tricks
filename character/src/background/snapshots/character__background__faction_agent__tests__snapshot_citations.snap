---
source: character/src/background/faction_agent.rs
expression: background.citations()

---
- - Scag
  - 147

